### PR TITLE
[pipelining] Add microbatch split and merge utils

### PR DESCRIPTION
**Pull request by kwen2501**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #125273
* #124958
* #124875
* #124776

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @chauhang @d4l3k

### CODE REVIEW

File: torch/distributed/pipelining/microbatch.py [hunk 1]
```diff
@@ -0,0 +1,420 @@
...
+    args_split = []
+    for chunk_args in args_split_dict:
+        args_split.append(tuple(chunk_args[i] for i in range(len(chunk_args))))
+
+    return args_split, kwargs_split
+
+
+def merge_chunks(
+    chunks: List[Any],
+    chunk_spec,
```
Comment (H-Huang) [inline torch/distributed/pipelining/microbatch.py:303]: missing type annotation

File: docs/source/distributed.pipelining.rst [hunk 2]
```diff
@@ -176,3 +176,16 @@ Note that since we split our model into three stages, we must run this script wi
...
   torchrun --nproc_per_node=3 example.py
+
+Microbatch Utilities
+====================
+
+.. automodule:: torch.distributed.pipelining.microbatch
+
+.. currentmodule:: torch.distributed.pipelining.microbatch
+
+.. autoclass:: TensorChunkSpec
```
Comment (H-Huang) [inline docs/source/distributed.pipelining.rst:187]: nit: May want to have block comment in class definition describing the class so it shows up in the docs.

**Review by H-Huang — APPROVED** [review]
Nice!

File: torch/distributed/pipelining/microbatch.py (hunk 1 above)
Comment (kwen2501) [reply torch/distributed/pipelining/microbatch.py:303]: Yeah, I will think about one there. That spec can really be Any.

**Pull request closed without merging**